src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -86,5 +113,258 @@ private protected IReadOnlyList<PackageReference> GetTransitivePackageReferences
+
+            if (cacheMissAssets || IsPackageSpecDifferent(currentPackageSpec, cachedPackageSpec))
+            {
+                _lastTimeAssetsModified = assets.LastWriteTimeUtc;
+                _lastPackageSpec = new WeakReference<PackageSpec>(currentPackageSpec);
+                IsInstalledAndTransitiveComputationNeeded = true;
+                IsTransitiveComputationNeeded = true;
+            }
+
+            return currentPackageSpec;
+        }
+
+        /// <summary>
+        /// Obtains targets section from project assets file (project.assets.json)
+        /// </summary>
+        /// <param name="ct">Cancellation token for async operation</param>
+        /// <returns>A list of dependencies, indexed by framework/RID</returns>
+        /// <remarks>Assets file reading occurs in a background thread</remarks>
+        /// <seealso cref="GetAssetsFilePathAsync"/>
+        protected async ValueTask<IList<LockFileTarget>> GetTargetsListAsync(CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+
+            await TaskScheduler.Default;
+
+            string assetsFilePath = await GetAssetsFilePathAsync();
+            LockFile lockFile = LockFileUtilities.GetLockFile(assetsFilePath, NullLogger.Instance);
+
+            return lockFile?.Targets;
+        }
+
+        /// <summary>
+        /// Runs Depth First Search recursively to mark current and dependend nodes with top dependency
+        /// </summary>
+        /// <param name="top">Top, Direct dependency</param>
+        /// <param name="current">Current package/node to visit</param>
+        /// <param name="graph">Package dependency graph, from assets file</param>
+        /// <param name="visited">Dictionary to remember visited nodes</param>
+        /// <param name="fxRidEntry">Framework/Runtime-ID associated with current <paramref name="graph"/></param>
+        private void MarkTransitiveOrigin(PackageReference top, PackageReference current, LockFileTarget graph, Dictionary<PackageIdentity, bool?> visited, FrameworkRuntimePair fxRidEntry, CancellationToken token)
+        {
+            token.ThrowIfCancellationRequested();
+
+            LockFileTargetLibrary node = graph
+                .Libraries
+                .Where(lib => string.Equals(lib.Name, current.PackageIdentity.Id, StringComparison.OrdinalIgnoreCase)
+                        && current.AllowedVersions.Satisfies(lib.Version)
+                        && lib.Type == LibraryType.Package.Value)
+                .FirstOrDefault();
+
+            if (node != default)
+            {
+                visited[current.PackageIdentity] = true; // visited
+
+                // Update cache
+                TransitiveEntry cachedEntry = GetCachedTransitiveOrigin(current.PackageIdentity);
+                if (cachedEntry == null)
+                {
+                    cachedEntry = new Dictionary<FrameworkRuntimePair, IList<PackageReference>>
```

Comment:
Could we replace with `cachedEntry = new Dictionary<TransitiveEntry>`... ?

Reply:
No, because Transitive entry is an interface type